### DISCUSSION TITLE
Routing Scope vs Resources

### DISCUSSION THREAD

**Discussion by felipero** [actix/actix-web#1908] [Q&A] [answered]
Hi guys, congrats on the framework. It is nice and fun to play with.

I have a question about the routing, that is not clear in the documentation. What's the practical difference between Scope and Resource in the routing. Why I can add services to a Scope but not to a Resource? I know a resource is registered into the ResourceMap, but looks like the Scope is too. So, I'm not sure what the real differences are, and I think that should be written in to the documentation. I'd be happy to send a PR if I figure this out.

**Comment by fakeshadow** [comment, marked as answer]
`Scope` has a private `Router` type which can route incoming request to it's children services.
`Resource` does not know anything about routing.

**Reply by felipero** [reply]
Thanks for the answer, I appreciate.

`Resource` do have a routes Vec. When you say they _don't know anything about routing_ are you saying they behave just like a route and not a router, while Scope behaves like a Router?

**Reply by fakeshadow** [reply]
Don't take the naming too literal.
`Resource` only match on `Guard` trait object and return the first one that does not conflict with the guards. 

Or to be specific. `Route` is a type of serivce that has associate guards like methods and headers. The routing part is done elsewhere

**Reply by felipero** [reply]
I was under the impression the `Resource` was routing, because of this code:
```rust
impl Service<ServiceRequest> for ResourceService {
    type Response = ServiceResponse;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    actix_service::always_ready!();

    fn call(&mut self, mut req: ServiceRequest) -> Self::Future {
        for route in self.routes.iter_mut() {
            if route.check(&mut req) {
                if let Some(ref app_data) = self.app_data {
                    req.add_data_container(app_data.clone());
                }
                return route.call(req);
            }
        }
        if let Some(ref app_data) = self.app_data {
            req.add_data_container(app_data.clone());
        }
        self.default.call(req)
    }
}
```
Guess I need to learn more about the actix-http and actix-service.

So, why can't `Resource` take a more generic serviceFactory, like `Scope` do? That's the origin of my confusion.

**Reply by fakeshadow** [reply]
It's designed not to. `Resource` is used to group a set of `Route` services and call them based on the `Guard` you define.
`Resource` or `Scope` are just used to construct some type impl `Service` trait. There is no functional difference between them and they behave differently because their state and service call logic.(Which is intended)

**Reply by felipero** [reply]
Makes sense. Thanks a lot for the answers, very helpful.
I'd love to contribute to the project, even if just with some documentation. Is there anything at the top of your head that I could help with?

**Reply by fakeshadow** [reply]
Thanks for asking. 

We have a lot of issues across the domain projects you may find interest into. 
The example repo also would like more examples to wide the coverage of different uses.
You can feel free to start new issue or PR if you find something thinking worth look into.
